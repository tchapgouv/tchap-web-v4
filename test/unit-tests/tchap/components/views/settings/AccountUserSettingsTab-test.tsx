import { render } from "@testing-library/react";
import React from "react";
import { MatrixClient } from "matrix-js-sdk/src/matrix";
import { logger } from "matrix-js-sdk/src/logger";
import { mocked, MockedObject } from "jest-mock";
import AccountUserSettingsTab from "matrix-react-sdk/src/components/views/settings/tabs/user/AccountUserSettingsTab";
import { SdkContextClass, SDKContext } from "matrix-react-sdk/src/contexts/SDKContext";
import SettingsStore from "matrix-react-sdk/src/settings/SettingsStore";
import { UIFeature } from "matrix-react-sdk/src/settings/UIFeature";
import { OidcClientStore } from "matrix-react-sdk/src/stores/oidc/OidcClientStore";
import MatrixClientContext from "matrix-react-sdk/src/contexts/MatrixClientContext";
import {
    getMockClientWithEventEmitter,
    mockClientMethodsServer,
    mockClientMethodsUser,
    mockPlatformPeg,
} from "matrix-react-sdk/test/test-utils";
import { useId } from "matrix-react-sdk/src/utils/useId";

jest.mock(
    "matrix-react-sdk/src/components/views/settings/ChangePassword",
    () =>
        ({ onError, onFinished }: { onError: (e: Error) => void; onFinished: () => void }) => {
            return <button>Mock change password</button>;
        },
);

jest.mock("matrix-react-sdk/src/utils/useId");

describe("<AccountUserSettingsTab />", () => {
    const defaultProps = {
        closeSettingsFn: jest.fn(),
    };

    const userId = "@alice:server.org";
    let mockClient: MockedObject<MatrixClient>;

    let stores: SdkContextClass;

    const getComponent = () => (
        <MatrixClientContext.Provider value={mockClient}>
            <SDKContext.Provider value={stores}>
                <AccountUserSettingsTab {...defaultProps} />
            </SDKContext.Provider>
        </MatrixClientContext.Provider>
    );

    beforeEach(() => {
        jest.spyOn(SettingsStore, "getValue").mockReturnValue(false);
        mockPlatformPeg();
        jest.clearAllMocks();
        jest.spyOn(SettingsStore, "getValue").mockRestore();
        jest.spyOn(logger, "error").mockRestore();

        mockClient = getMockClientWithEventEmitter({
            ...mockClientMethodsUser(userId),
            ...mockClientMethodsServer(),
            getCapabilities: jest.fn(),
            getThreePids: jest.fn(),
            getIdentityServerUrl: jest.fn(),
            deleteThreePid: jest.fn(),
        });

        mockClient.getCapabilities.mockResolvedValue({});
        mockClient.getThreePids.mockResolvedValue({
            threepids: [],
        });
        mockClient.deleteThreePid.mockResolvedValue({
            id_server_unbind_result: "success",
        });

        stores = new SdkContextClass();
        stores.client = mockClient;
        // stub out this store completely to avoid mocking initialisation
        const mockOidcClientStore = {} as unknown as OidcClientStore;
        jest.spyOn(stores, "oidcClientStore", "get").mockReturnValue(mockOidcClientStore);
        jest.spyOn(SettingsStore, "getValue").mockImplementation(
            (settingName: string) => settingName === UIFeature.Deactivate,
        );

        // labelled id on username is autogeneraed, we fix it so that snapshot is always the same
        mocked(useId).mockImplementation(() => "1234");
    });

    afterEach(() => {
        jest.restoreAllMocks();
    });

    describe("common view snapshot", () => {
        it("should render section when account deactivation feature is enabled", () => {
            const { container } = render(getComponent());

            expect(container).toMatchSnapshot();
        });
    });
});
