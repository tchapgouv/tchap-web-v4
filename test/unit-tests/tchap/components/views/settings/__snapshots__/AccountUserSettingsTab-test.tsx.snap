// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AccountUserSettingsTab /> common view snapshot should render section when account deactivation feature is enabled 1`] = `
<div>
  <div
    class="mx_SettingsTab"
    data-testid="mx_AccountUserSettingsTab"
  >
    <div
      class="mx_SettingsTab_sections"
    >
      <div
        class="mx_SettingsSection"
      >
        <div
          class="mx_SettingsSection_subSections"
        >
          <div
            class="mx_UserProfileSettings"
          >
            <h2>
              Profile
            </h2>
            <div>
              Your account is managed separately by an identity provider and so some of your personal information can’t be changed here.
            </div>
            <div
              class="mx_UserProfileSettings_profile"
            >
              <div
                aria-label="Profile picture"
                class="mx_AvatarSetting_avatar"
                role="group"
              >
                <div
                  aria-disabled="true"
                  class="mx_AccessibleButton mx_AvatarSetting_avatarPlaceholder mx_AvatarSetting_avatarDisplay mx_AccessibleButton_disabled"
                  disabled=""
                  role="button"
                  tabindex="-1"
                >
                  <span
                    class="_avatar_mcap2_17 mx_BaseAvatar _avatar-imageless_mcap2_61"
                    data-color="2"
                    data-testid="avatar-img"
                    data-type="round"
                    role="presentation"
                    style="--cpd-avatar-size: 90px;"
                  >
                    
                  </span>
                </div>
              </div>
              <form
                class="_root_dgy0u_24 mx_UserProfileSettings_profile_displayName"
              >
                <div
                  class="_field_dgy0u_34"
                >
                  <label
                    class="_label_dgy0u_67"
                    for="radix-0"
                  >
                    Display Name
                  </label>
                  <div
                    class="_controls_1h4nb_17"
                  >
                    <input
                      class="_control_9gon8_18"
                      disabled=""
                      id="radix-0"
                      name="input"
                      title=""
                      value=""
                    />
                  </div>
                </div>
              </form>
            </div>
            <div
              class="mx_UserProfileSettings_profile_controls_userId"
            >
              <div
                class="mx_UserProfileSettings_profile_controls_userId_label"
                id="1234"
              >
                Username
              </div>
              <div
                aria-labelledby="1234"
                class="mx_CopyableText mx_CopyableText_border"
              >
                @alice:server.org
                <div
                  aria-label="Copy"
                  class="mx_AccessibleButton mx_CopyableText_copyButton"
                  role="button"
                  tabindex="0"
                />
              </div>
            </div>
            <div
              class="mx_Flex mx_UserProfileSettings_profile_buttons"
              style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: start; --mx-flex-justify: start; --mx-flex-gap: var(--cpd-space-4x);"
            >
              <div
                class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_danger_outline"
                role="button"
                tabindex="0"
              >
                <div
                  class="mx_UserProfileSettings_accountmanageIcon"
                  height="24"
                  width="24"
                />
                Sign out
              </div>
            </div>
          </div>
          <div
            class="mx_SettingsSubsection"
          >
            <div
              class="mx_SettingsSubsectionHeading"
            >
              <h3
                class="mx_Heading_h4 mx_SettingsSubsectionHeading_heading"
              >
                settings
              </h3>
            </div>
            <div
              class="mx_SettingsSubsection_content mx_SettingsSubsection_contentStretch"
            >
              <div
                class="mx_SettingsFlag"
              >
                <label
                  class="mx_SettingsFlag_label"
                  for="redList-toggle"
                >
                  <span
                    class="mx_SettingsFlag_labelText"
                  >
                    settings
                  </span>
                  <div
                    class="mx_SettingsFlag_microcopy"
                  >
                    settings
                  </div>
                </label>
                <div
                  aria-checked="false"
                  aria-disabled="false"
                  aria-label="settings"
                  class="mx_AccessibleButton mx_ToggleSwitch mx_ToggleSwitch_enabled"
                  id="redList-toggle"
                  role="switch"
                  tabindex="0"
                >
                  <div
                    class="mx_ToggleSwitch_ball"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="mx_SettingsSection"
      >
        <h2
          class="mx_Heading_h3"
        >
          Deactivate Account
        </h2>
        <div
          class="mx_SettingsSection_subSections"
        >
          <div
            class="mx_SettingsSubsection"
            data-testid="account-management-section"
          >
            <div
              class="mx_SettingsSubsectionHeading"
            >
              <h3
                class="mx_Heading_h4 mx_SettingsSubsectionHeading_heading"
              >
                Account management
              </h3>
            </div>
            <div
              class="mx_SettingsSubsection_description"
            >
              <div
                class="mx_SettingsSubsection_text"
              >
                Deactivating your account is a permanent action — be careful!
              </div>
            </div>
            <div
              class="mx_SettingsSubsection_content"
            >
              <div
                class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_danger"
                role="button"
                tabindex="0"
              >
                Deactivate Account
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
